File: swift/Sources/MLXAudioModels/Banquet/BandSplit.swift
// BandSplit.swift
// Band splitting module for Banquet query-based source separation.

import Foundation
import MLX
import MLXAudioPrimitives
import MLXNN

/// Normalization + Fully Connected layer for band processing.
///
/// Applies LayerNorm over flattened band input, then projects to embedding dimension.
public class NormFC: Module, @unchecked Sendable {
    @ModuleInfo var norm: LayerNorm
    @ModuleInfo var fc: Linear

    let treatChannelAsFeature: Bool

    /// Creates a NormFC layer.
    ///
    /// - Parameters:
    ///   - embDim: Embedding dimension output
    ///   - bandwidth: Width of frequency band in bins
    ///   - inChannel: Number of input channels
    ///   - treatChannelAsFeature: Whether to treat channels as features (default: true)
    public init(
        embDim: Int,
        bandwidth: Int,
        inChannel: Int,
        treatChannelAsFeature: Bool = true
    ) {
        self.treatChannelAsFeature = treatChannelAsFeature
        let reim = 2  // Real and imaginary parts

        // LayerNorm over flattened input
        _norm.wrappedValue = LayerNorm(dimensions: inChannel * bandwidth * reim)

        // FC input dimension depends on treatChannelAsFeature
        var fcIn = bandwidth * reim
        var outputDim = embDim
        if treatChannelAsFeature {
            fcIn *= inChannel
        } else {
            precondition(embDim % inChannel == 0, "embDim must be divisible by inChannel")
            outputDim = embDim / inChannel
        }

        _fc.wrappedValue = Linear(fcIn, outputDim)
    }

    /// Forward pass.
    ///
    /// - Parameter xb: Band input [batch, n_time, in_chan, reim * bandwidth]
    /// - Returns: Band embeddings [batch, n_time, emb_dim]
    public func callAsFunction(_ xb: MLXArray) -> MLXArray {
        let shape = xb.shape
        let batch = shape[0]
        let nTime = shape[1]
        let inChan = shape[2]
        let ribw = shape[3]

        // Flatten for LayerNorm: [batch, n_time, in_chan * reim * bandwidth]
        var x = xb.reshaped([batch, nTime, inChan * ribw])
        x = norm(x)

        if !treatChannelAsFeature {
            // Reshape back: [batch, n_time, in_chan, reim * bandwidth]
            x = x.reshaped([batch, nTime, inChan, ribw])
        }

        // Project to embedding dimension
        var z = fc(x)

        if !treatChannelAsFeature {
            // Flatten channel dimension into embedding
            let zShape = z.shape
            z = z.reshaped([zShape[0], zShape[1], zShape[2] * zShape[3]])
        }

        return z  // [batch, n_time, emb_dim]
    }
}

/// Band splitting module for frequency decomposition.
///
/// Splits complex spectrogram into frequency bands and projects each band
/// to an embedding dimension using per-band NormFC modules.
public class BandSplitModule: Module, @unchecked Sendable {
    @ModuleInfo(key: "norm_fc_modules") var normFCModules: [NormFC]

    let bandSpecs: [BandSpec]
    let bandWidths: [Int]
    let nBands: Int
    let embDim: Int

    /// Creates a BandSplitModule.
    ///
    /// - Parameters:
    ///   - inChannel: Number of input audio channels
    ///   - bandSpecs: List of band specifications (start_bin, end_bin)
    ///   - embDim: Embedding dimension for band features
    ///   - treatChannelAsFeature: Whether to treat channels as features
    public init(
        inChannel: Int,
        bandSpecs: [BandSpec],
        embDim: Int = 128,
        treatChannelAsFeature: Bool = true
    ) {
        self.bandSpecs = bandSpecs
        self.bandWidths = bandSpecs.map { $0.bandwidth }
        self.nBands = bandSpecs.count
        self.embDim = embDim

        // Create per-band NormFC modules
        _normFCModules.wrappedValue = bandWidths.map { bw in
            NormFC(
                embDim: embDim,
                bandwidth: bw,
                inChannel: inChannel,
                treatChannelAsFeature: treatChannelAsFeature
            )
        }
    }

    /// Forward pass.
    ///
    /// - Parameter x: Complex spectrogram [batch, in_chan, n_freq, n_time]
    /// - Returns: Band embeddings [batch, n_bands, n_time, emb_dim]
    public func callAsFunction(_ x: ComplexArray) -> MLXArray {
        let shape = x.shape
        let batch = shape[0]
        let inChan = shape[1]
        let nTime = shape[3]

        // Convert complex to real/imag: [batch, in_chan, n_freq, n_time, 2]
        let xr = MLX.stacked([x.real, x.imag], axis: -1)

        // Permute: [batch, n_time, in_chan, 2, n_freq]
        let xrT = xr.transposed(0, 3, 1, 4, 2)

        // Process each band
        var outputs: [MLXArray] = []
        for (i, nfm) in normFCModules.enumerated() {
            let fstart = bandSpecs[i].startBin
            let fend = bandSpecs[i].endBin

            // Extract band: [batch, n_time, in_chan, reim, bandwidth]
            let xb = xrT[0..., 0..., 0..., 0..., fstart..<fend]

            // Flatten reim and bandwidth: [batch, n_time, in_chan, reim * bandwidth]
            let xbShape = xb.shape
            let xbFlat = xb.reshaped([xbShape[0], xbShape[1], xbShape[2], xbShape[3] * xbShape[4]])

            // Apply NormFC
            let zb = nfm(xbFlat)  // [batch, n_time, emb_dim]
            outputs.append(zb)
        }

        // Stack bands: [batch, n_bands, n_time, emb_dim]
        let z = MLX.stacked(outputs, axis: 1)

        return z
    }

    /// Creates a BandSplitModule from configuration.
    ///
    /// - Parameters:
    ///   - config: Banquet configuration
    /// - Returns: Configured BandSplitModule
    public static func fromConfig(_ config: BanquetConfig) -> BandSplitModule {
        let bandSpecs = MusicalBandsplitSpecification.generate(
            nFFT: config.nFFT,
            sampleRate: config.sampleRate,
            nBands: config.nBands
        )

        return BandSplitModule(
            inChannel: config.inChannel,
            bandSpecs: bandSpecs,
            embDim: config.embDim
        )
    }
}

/// Musical band specification generator.
///
/// Generates frequency bands using MIDI note spacing for perceptually uniform bands.
public struct MusicalBandsplitSpecification {

    /// Generates band specifications for the given parameters.
    ///
    /// - Parameters:
    ///   - nFFT: FFT size
    ///   - sampleRate: Sample rate in Hz
    ///   - nBands: Number of frequency bands
    ///   - fMin: Minimum frequency (default: sampleRate / nFFT)
    ///   - fMax: Maximum frequency (default: sampleRate / 2)
    /// - Returns: Array of BandSpec
    public static func generate(
        nFFT: Int,
        sampleRate: Int,
        nBands: Int = 64,
        fMin: Float? = nil,
        fMax: Float? = nil
    ) -> [BandSpec] {
        let fs = Float(sampleRate)
        let nFreqs = nFFT / 2 + 1
        let df = fs / Float(nFFT)

        let fMaxVal = fMax ?? fs / 2
        let fMinVal = fMin ?? fs / Float(nFFT)

        // Calculate octave spacing
        let nOctaves = log2(fMaxVal / fMinVal)
        let nOctavesPerBand = nOctaves / Float(nBands)
        let bandwidthMult = pow(2.0, nOctavesPerBand)

        // Convert to MIDI for linear spacing
        let lowMidi = max(0, hzToMidi(fMinVal))
        let highMidi = hzToMidi(fMaxVal)

        var midiPoints: [Float] = []
        for i in 0..<nBands {
            let t = Float(i) / Float(nBands - 1)
            midiPoints.append(lowMidi + t * (highMidi - lowMidi))
        }

        let hzPts = midiPoints.map { midiToHz($0) }

        // Calculate band boundaries
        let lowPts = hzPts.map { $0 / bandwidthMult }
        let highPts = hzPts.map { $0 * bandwidthMult }

        let lowBins = lowPts.map { Int(floor($0 / df)) }
        let highBins = highPts.map { Int(ceil($0 / df)) }

        // Create filterbank mask
        var fb = [[Float]](repeating: [Float](repeating: 0, count: nFreqs), count: nBands)
        for i in 0..<nBands {
            let start = max(0, lowBins[i])
            let end = min(nFreqs, highBins[i] + 1)
            for j in start..<end {
                fb[i][j] = 1.0
            }
        }

        // Extend first and last bands
        for j in 0..<max(0, lowBins[0]) {
            fb[0][j] = 1.0
        }
        for j in min(nFreqs, highBins[nBands - 1] + 1)..<nFreqs {
            fb[nBands - 1][j] = 1.0
        }

        // Extract band specs
        var bandSpecs: [BandSpec] = []
        for i in 0..<nBands {
            var startIdx: Int?
            var endIdx: Int?
            for j in 0..<nFreqs {
                if fb[i][j] > 0 {
                    if startIdx == nil {
                        startIdx = j
                    }
                    endIdx = j + 1
                }
            }
            if let start = startIdx, let end = endIdx {
                bandSpecs.append(BandSpec(startBin: start, endBin: end))
            }
        }

        return bandSpecs
    }

    /// Convert frequency in Hz to MIDI note number.
    private static func hzToMidi(_ hz: Float) -> Float {
        return 12.0 * log2(hz / 440.0) + 69.0
    }

    /// Convert MIDI note number to frequency in Hz.
    private static func midiToHz(_ midi: Float) -> Float {
        return 440.0 * pow(2.0, (midi - 69.0) / 12.0)
    }
}

File: python/mlx_audio/models/banquet/bandsplit.py
"""BandSplit module for Banquet model."""

from __future__ import annotations

import mlx.core as mx
import mlx.nn as nn

from mlx_audio.models.banquet.utils import (
    MusicalBandsplitSpecification,
    band_widths_from_specs,
    check_no_gap,
    check_no_overlap,
    check_nonzero_bandwidth,
)


class NormFC(nn.Module):
    """Normalization + Fully Connected layer for band processing.

    Applies LayerNorm over flattened band input, then projects to embedding dimension.
    """

    def __init__(
        self,
        emb_dim: int,
        bandwidth: int,
        in_channel: int,
        normalize_channel_independently: bool = False,
        treat_channel_as_feature: bool = True,
    ) -> None:
        super().__init__()

        if normalize_channel_independently:
            raise NotImplementedError("normalize_channel_independently not supported")

        self.treat_channel_as_feature = treat_channel_as_feature
        reim = 2  # Real and imaginary parts

        # LayerNorm over flattened input
        self.norm = nn.LayerNorm(in_channel * bandwidth * reim)

        # FC input dimension depends on treat_channel_as_feature
        fc_in = bandwidth * reim
        if treat_channel_as_feature:
            fc_in *= in_channel
        else:
            assert emb_dim % in_channel == 0
            emb_dim = emb_dim // in_channel

        self.fc = nn.Linear(fc_in, emb_dim)

    def __call__(self, xb: mx.array) -> mx.array:
        """Forward pass.

        Args:
            xb: Band input [batch, n_time, in_chan, reim * bandwidth]

        Returns:
            Band embeddings [batch, n_time, emb_dim]
        """
        batch, n_time, in_chan, ribw = xb.shape

        # Flatten for LayerNorm: [batch, n_time, in_chan * reim * bandwidth]
        xb = self.norm(xb.reshape(batch, n_time, in_chan * ribw))

        if not self.treat_channel_as_feature:
            # Reshape back: [batch, n_time, in_chan, reim * bandwidth]
            xb = xb.reshape(batch, n_time, in_chan, ribw)

        # Project to embedding dimension
        zb = self.fc(xb)

        if not self.treat_channel_as_feature:
            # Flatten channel dimension into embedding
            batch, n_time, in_chan, emb_dim_per_chan = zb.shape
            zb = zb.reshape(batch, n_time, in_chan * emb_dim_per_chan)

        return zb  # [batch, n_time, emb_dim]


class BandSplitModule(nn.Module):
    """Band splitting module for frequency decomposition.

    Splits complex spectrogram into frequency bands and projects each band
    to an embedding dimension using per-band NormFC modules.
    """

    def __init__(
        self,
        in_channel: int,
        band_specs: list[tuple[int, int]],
        emb_dim: int = 128,
        require_no_overlap: bool = False,
        require_no_gap: bool = True,
        normalize_channel_independently: bool = False,
        treat_channel_as_feature: bool = True,
    ) -> None:
        super().__init__()

        check_nonzero_bandwidth(band_specs)

        if require_no_gap:
            check_no_gap(band_specs)

        if require_no_overlap:
            check_no_overlap(band_specs)

        self.band_specs = band_specs
        self.band_widths = band_widths_from_specs(band_specs)
        self.n_bands = len(band_specs)
        self.emb_dim = emb_dim

        # Create per-band NormFC modules
        self.norm_fc_modules = [
            NormFC(
                emb_dim=emb_dim,
                bandwidth=bw,
                in_channel=in_channel,
                normalize_channel_independently=normalize_channel_independently,
                treat_channel_as_feature=treat_channel_as_feature,
            )
            for bw in self.band_widths
        ]

    def __call__(self, x: mx.array) -> mx.array:
        """Forward pass.

        Args:
            x: Complex spectrogram [batch, in_chan, n_freq, n_time]
               Note: x should be complex-valued

        Returns:
            Band embeddings [batch, n_bands, n_time, emb_dim]
        """
        batch, in_chan, n_freq, n_time = x.shape

        # Convert complex to real/imag: [batch, in_chan, n_freq, n_time, 2]
        # If x is already real with last dim 2, use as-is
        if x.dtype in (mx.complex64,):
            xr = mx.stack([x.real, x.imag], axis=-1)
        else:
            # Assume x is [batch, in_chan, n_freq, n_time] with interleaved or
            # already has trailing dim
            xr = mx.stack([x.real, x.imag], axis=-1)

        # Permute: [batch, n_time, in_chan, 2, n_freq]
        xr = mx.transpose(xr, (0, 3, 1, 4, 2))

        batch, n_time, in_chan, reim, _ = xr.shape

        # Initialize output
        z = mx.zeros((batch, self.n_bands, n_time, self.emb_dim))

        # Process each band
        outputs = []
        for i, nfm in enumerate(self.norm_fc_modules):
            fstart, fend = self.band_specs[i]
            # Extract band: [batch, n_time, in_chan, reim, bandwidth]
            xb = xr[..., fstart:fend]
            # Flatten reim and bandwidth: [batch, n_time, in_chan, reim * bandwidth]
            xb = xb.reshape(batch, n_time, in_chan, -1)
            # Apply NormFC
            zb = nfm(xb)  # [batch, n_time, emb_dim]
            outputs.append(zb)

        # Stack bands: [batch, n_bands, n_time, emb_dim]
        z = mx.stack(outputs, axis=1)

        return z

    @classmethod
    def from_config(
        cls,
        n_fft: int = 2048,
        sample_rate: int = 44100,
        n_bands: int = 64,
        in_channel: int = 2,
        emb_dim: int = 128,
        band_type: str = "musical",
        **kwargs,
    ) -> BandSplitModule:
        """Create BandSplitModule from configuration.

        Args:
            n_fft: FFT size
            sample_rate: Sample rate in Hz
            n_bands: Number of frequency bands
            in_channel: Number of audio channels
            emb_dim: Embedding dimension
            band_type: Band specification type ("musical")
            **kwargs: Additional arguments passed to __init__

        Returns:
            Configured BandSplitModule
        """
        if band_type != "musical":
            raise ValueError(f"Unsupported band_type: {band_type}")

        spec = MusicalBandsplitSpecification(
            nfft=n_fft,
            fs=sample_rate,
            n_bands=n_bands,
        )

        return cls(
            in_channel=in_channel,
            band_specs=spec.get_band_specs(),
            emb_dim=emb_dim,
            **kwargs,
        )
